### DISCUSSION TITLE
Swift GlobalActors & Dependencies

### DISCUSSION THREAD

**Discussion by idelfonsog2** [pointfreeco/swift-dependencies#36] [Q&A] [answered]
I know the topic of working with Swift Actors was mentioned at some point during the TCA video series. I'm curious to know if there are any reasons **_not_** to mix Dependency with a GlobalActor type.

The GlobalActor type enhances the Singleton pattern we have. A dependency can have a mutable state. I did read the docs about [ActorIsolated](https://pointfreeco.github.io/swift-dependencies/main/documentation/dependencies/actorisolated) Type in this library.

After implementing the globalActor to one of my Dependency wrapper, it's making me think maybe another need the Dependency var

**Comment by tgrapperon** [comment, marked as answer]
Having a global actor would help dependencies isolated on the same actor to interdepend synchronously. As I can imagine this happening in user code bases, I'm not sure there is something the library can do to help in this area, or maybe I'm not understanding your suggestion correctly.

You can of course isolate parts of your dependencies to global actors, and this is "commonly" performed with `MainActor`. It is also possible to completely isolate dependencies, but you'll need to mark `liveValue` & co as `nonisolated` to conform to `DependencyKey` (if you're using `Self` as `Value`), and you'll need to do a little gymnastic to synchronously create values of your isolated dependency to conform to the protocol anyway.

If this is not a singleton, I wouldn't recommend to implement your dependency as an actor itself, global or not. If you're not cautious, this reference-type will leak through dependencies overrides, and you may find bugs caused by action at distance that are difficult to fix. Whereas you're making this singleton a global actor is up to you, but I'm not exactly seeing how the library could help here, beside providing helpers to handle this object synchronously (if this is even possible in the general picture). Values that you extract from `@Dependency` can already be isolated to a global actor.

Do you have an example where the global actor is causing friction, or where the situation could be improved?

**Reply by idelfonsog2** [reply]
Thank you for the response @tgrapperon 

I wasn't sure to how (or if at all) did the dependency library could help enhance or mitigate issues with we work with Swift structure concurrency.

My example is the following:

- I have a wrapper around an Analytics SDK
- I use the instance methods across `Tasks`.
- Beside calling the SDK getters it also modifies the properties of this type wrapper instance

**Reply by idelfonsog2** [reply]
After thinking it for a day, I went with the following outcome:
- Define my GlobalActor
- Wrapped my GlobalActor in Dependency Struct and offer it to the environment:

```swift
@globalActor actor AnalyticsSingleton: Analytics {...}

struct AnalyticsClient: Analytics {
  func checkStatus() async throws {
    AnalyticsSingleton.shared.checkStatus()
  }
} // uses AnalyticsSingleton
struct AnalyticsMockClient: Analytics {...}

private enum AnalyticsConfigKey: DependencyKey {
  public static let liveValue: Analytics = AnalyticsClient() // ✅
  public static let testValue: Analytics = AnalyticsMockClient() // 🤡
}
```

**Reply by tgrapperon** [reply]
Thanks for the follow up. The thing that eludes me is the global actor! I feel that this would work the same with a simple actor having a `shared` static variable (the singleton). As I'm understanding it, the real interest would be if your dependency's clients are themself marked as `@AnalyticsSingleton`. This would allow them to work as if they were the actor themselves, that is to interact synchronously with actor-isolated properties and functions.
Maybe you're conflating being a global actor and being an actor with a shared instance.
Your problem can modelize very well using a shared actor instance. As an aside, you can probably refactor your dependency in the style described [here](https://pointfreeco.github.io/swift-dependencies/main/documentation/dependencies/designingdependencies/) and avoid having to create a protocol and type for the mock.
But going back to the main question: is there anything that you do with your dependency that you couldn't do by simply removing the `@globalActor` annotation? If it still builds after having removed this annotation, then you can probably go without a global actor. Please let me know if I missed your use case.

**Reply by stephencelis** [reply]
@idelfonsog2 I would either avoid custom global actors for now or isolate them to the "live" branches of your dependency code. I believe global actors perform their work on their own executors that run outside the default and main executors, and Swift doesn't appear to provide the means to override this kind of executor hook yet, which means any code that tries to control the default and main executors for the purpose of reliable testing will not work for a custom global actor. While we haven't employed these kinds of executor overrides in our tooling yet, we do have some experiments that we employ in our library unit tests that are promising, and we may incorporate them into library tools in the future.

**Reply by idelfonsog2** [reply]
Thank you 🙏🏻 @tgrapperon @stephencelis

**Reply by idelfonsog2** [reply]
> Thanks for the follow up. The thing that eludes me is the global actor! I feel that this would work the same with a simple actor having a `shared` static variable (the singleton). As I'm understanding it, the real interest would be if your dependency's clients are themself marked as `@AnalyticsSingleton`. This would allow them to work as if they were the actor themselves, that is to interact synchronously with actor-isolated properties and functions.
> Maybe you're conflating being a global actor and being an actor with a shared instance.
> Your problem can modelize very well using a shared actor instance. As an aside, you can probably refactor your dependency in the style described [here](https://pointfreeco.github.io/swift-dependencies/main/documentation/dependencies/designingdependencies/) and avoid having to create a protocol and type for the mock.
> But going back to the main question: is there anything that you do with your dependency that you couldn't do by simply removing the `@globalActor` annotation? If it still builds after having removed this annotation, then you can probably go without a global actor. Please let me know if I missed your use case.

I do! My global actor holds a reference to X third-party SDK properties that I'm instantiating in the initializer of my globalActor. Any code in the application should interact with this unique instance because the configuration in the `init` needs to happen in the ... AppDelegate `appDidFinishLunching` method. Thats why Im mainy trying to enforce the use of `@globalActor`

```swift
import ThirdPartySDKAnalytics

@globalActor .... {
...
public init() {
  listener = AnalyticsListener(configPath:...)
}

func checkGuest() async {
  listener.remove()
  ...
}
```

**Reply by nashysolutions** [reply]
It might be good to put this in the documentation. Had memory leak issue today.

**Reply by mbrandonw** [reply]
Hi @nashysolutions, can you explain a bit more about what you want documented and the leak? I don't see anything in this thread that would cause leaks.

**Reply by nashysolutions** [reply]
Hi @mbrandonw sorry for late reply. @tgrapperon mentioned the following above

> ...If this is not a singleton, I wouldn't recommend to implement your dependency as an actor itself, global or not. If you're not cautious, this reference-type will leak through dependencies overrides...

**Reply by tgrapperon** [reply]
To be clear, the "leak" I was talking about wasn't a memory leak, but rather a way to say that the same instance may be shared by different override contexts simultaneously. There are thus cases where the parent may act using some child's overrides rather than its own values. 

For example, if your dependency is a class with configurable/unimplemented endpoints, if you override one in a child, this will also override the parent's one too, and this will furthermore not be reset when the child context ends.

Sorry if the term led to confusion. I should probably have used "bleed" instead.
